lib/spack/spack/installer.py

```diff
@@ -901,6 +919,13 @@ def __init__(
         # The initial start time for processing the spec
         self.start = start
 
+        # Make sure installed is an expected type
+        if isinstance(installed, (tuple, list)):
+            installed = set(installed)
+
+        if not isinstance(installed, set):
+            raise TypeError(f"{installed} is not a valid installed type")
```

Comment:
Can we get clearer error message? We have lots of things called `installed` in various parts of the code base.

What about:

```suggestion
            raise TypeError(f"'installed' argument to {type(self)}.__init__() must be a 'set'. Received {type(installed)}")
```

Reply:
I simplified your suggestion a bit.